Add memory_format support `to` and `type` operators

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27979 Add memory format support to `resize_as_` operator
* #27891 Cleanup testing of _like operators
* #27890 Add memory format support to `randn_like` operator
* #27889 Add memory format support to `randint_like` operator
* #27562 Add memory format support to `zeros_like` operator
* #27561 Add memory format support to `rand_like` operator
* #27270 Add memory format support to `ones_like` operator
* #27262 Add memory format support to `full_like` operator
* #27244 Add memory format support to `empty_like` operator
* #27228 Add memory format support to typecasting shortcuts `byte`,`char`,`double`,`bool`,`half`,`int`,`long`,`short`,`float`,`bfloat16`
* #27223 Add memory format support to `cpu` and `cuda` operators
* **#27107 Add memory_format support `to` and `type` operators**

Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17931062](https://our.internmc.facebook.com/intern/diff/D17931062)

VitalyFedyunin committed:
[WIP] Add memory_format support `to` and `type` operators

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Add memory_format support `to` and `type` operators"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Add memory_format support `to` and `type` operators"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Add memory_format support `to` and `type` operators"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory_format support `to` and `type` operators"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory_format support `to` and `type` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory_format support `to` and `type` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

ezyang commented:
> If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.

Define what it means to be in 'channels last' format? :)

ezyang commented on aten/src/ATen/core/TensorBody.h:
```diff
@@ -781,10 +781,10 @@ class CAFFE2_API Tensor {
-  Tensor to(Device device, ScalarType dtype, bool non_blocking=false, bool copy=false) const;
-  Tensor to(ScalarType dtype, bool non_blocking=false, bool copy=false) const;
-  Tensor to(const Tensor & other, bool non_blocking=false, bool copy=false) const;
+  Tensor to(const TensorOptions & options, bool non_blocking=false, bool copy=false, c10::optional<MemoryFormat> memory_format=c10::nullopt) const;
```
I told you this before, and I'm going to complain about it again here: MemoryFormat should be a parameter in TensorOptions, not a freestanding positional argument as it is here. The state of positional arguments here is basically unusable for actual users of this API.

VitalyFedyunin replied:
Going to be merged into TensorOptions as soon as TensorOptions codegen rewrite finished ( CC @izdeby )

ezyang commented on torch/csrc/autograd/utils/python_arg_parsing.h:
```diff
@@ -9,23 +9,23 @@ namespace torch { namespace autograd { namespace utils {
-    "to(Tensor tensor, bool non_blocking=False, bool copy=False)",
+    "to(Device device=None, ScalarType dtype=None, bool non_blocking=False, bool copy=False, *, MemoryFormat? memory_format=None)",
+    "to(ScalarType dtype, bool non_blocking=False, bool copy=False, *, MemoryFormat? memory_format=None)",
+    "to(Tensor tensor, bool non_blocking=False, bool copy=False, *, MemoryFormat? memory_format=None)",
```
Ugh, we should have made `non_blocking` and `copy` keyword only too XD

VitalyFedyunin replied:
Going to be separate PR.

ezyang commented on native_functions_test_metadata.yaml:
```diff
@@ -0,0 +1,678 @@
+- func: __and__
```
What's this?

VitalyFedyunin replied:
Not supposed to be part of PR. Removing.

ezyang commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -19,13 +19,25 @@ static inline Device ensure_has_index(Device device) {
-  auto r = at::empty(self.sizes(), options);
+static inline Tensor to_impl(const Tensor& self, const TensorOptions& options, bool non_blocking, c10::optional<c10::MemoryFormat> optional_memory_format) {
+  auto memory_format =
+      optional_memory_format.value_or(MemoryFormat::Contiguous);
```
This one will change later right?

VitalyFedyunin replied:
Yes, now we need to introduce additional argument and change codebase to use `.to(..., memory_format=contig)` wherever  developers counting on old `.to` side-effect.

ezyang commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -42,7 +54,9 @@ Tensor to(const Tensor& self, const TensorOptions& options, bool non_blocking, b
-      (!dtype_opt  || self.dtype()  ==  dtype_opt.value()) && !copy) {
+      (!dtype_opt || self.dtype() == dtype_opt.value()) && !copy &&
+      optional_memory_format.value_or(MemoryFormat::Preserve) ==
+          MemoryFormat::Preserve) {
```
Is this the only condition we should return self? Shouldn't we also return self if the requested memory format matches what the tensor is?

ezyang commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -52,33 +66,44 @@ Tensor to(const Tensor& self, const TensorOptions& options, bool non_blocking, b
-  if (self_options == options && !copy) {
+  if (self_options == options && !copy &&
+      optional_memory_format.value_or(MemoryFormat::Preserve) ==
+          MemoryFormat::Preserve) {
```
Seeing that this logic is replicated many times, maybe it is worth pushing it into the `to_impl` implementation...

ezyang commented:
Please rerequest review when this is ready

VitalyFedyunin committed:
Update on "Add memory_format support `to` and `type` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin commented:
> Define what it means to be in 'channels last' format? :)

https://github.com/pytorch/pytorch/blob/d70f8dd9646909f4d465dfc6a5094a63e83c7c3f/c10/core/TensorImpl.h#L1643

VitalyFedyunin committed:
Update on "Add memory_format support `to` and `type` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Fix merge conflict on "Add memory_format support `to` and `type` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Fix .to copy=True path on "Add memory_format support `to` and `type` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -19,13 +19,32 @@ static inline Device ensure_has_index(Device device) {
+  auto memory_format =
+      optional_memory_format.value_or(MemoryFormat::Contiguous);
+
+  if (self.options() == options && !copy &&
```
Don't compare options. See https://github.com/pytorch/pytorch/issues/25478 and  https://github.com/pytorch/pytorch/pull/27572 (btw, bikeshedding on TensorAxes welcome)

ezyang replied:
Result of in person conversation: Let's add something like "match TensorOptions with Tensor". This tests if a Tensor is "compatible" with a TensorOptions. (This is not an equality test, because options can be left unspecified, and in that case you always match.)

ezyang commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -41,44 +60,34 @@ Tensor to(const Tensor& self, const TensorOptions& options, bool non_blocking, b
-  return to_impl(self, self.options().device(device).dtype(dtype), non_blocking);
+  return to_impl(
+      self,
+      self.options().device(device).dtype(dtype),
```
I know this code originally existed, but I want to point out that it is pretty weird that we have to call `self.options()` here, because the point of TensorOptions is that it has optional fields, so `to_impl` should be the method that actually handles the population in the end. No change necessary.

ezyang reviewed (approved):
I did not carefully check the Python arg parser changes

ezyang commented:
To clarify on TensorOptions equality testing: I see in other parts of the code that this was a preexisting pattern. So I won't require it to be fixed in this PR.

VitalyFedyunin commented:
@ailzhang breaks XLA (as almost every PR in this stack)
